File: web/src/library.js
// Parse + validate a profile's tags.csv (mirrors engine/library.py). Loud on
// unknown vocab, bad tiers, duplicates, or empties — the library is the source
// of truth, so we stop here if it's wrong.

import { parseCSV } from "./csv.js";
import { normalizeIndustry, normalizeType } from "./vocab.js";

const REQUIRED = ["file", "name", "industries", "types", "tier"];
const TRUE = new Set(["true", "1", "yes", "y", "t"]);
const FALSE = new Set(["false", "0", "no", "n", "f", ""]);

function splitList(raw) {
  return (raw || "").split("|").map((s) => s.trim()).filter(Boolean);
}

export function parseLibrary(csvText) {
  const { header, records } = parseCSV(csvText);
  const missing = REQUIRED.filter((c) => !header.includes(c));
  if (missing.length) throw new Error(`tags.csv missing column(s): ${missing.join(", ")}`);

  const entries = [];
  const seen = new Set();
  records.forEach((row, idx) => {
    const line = idx + 2;
    const file = (row.file || "").trim();
    if (!file) throw new Error(`tags.csv line ${line}: empty 'file'`);
    if (seen.has(file)) throw new Error(`tags.csv line ${line}: duplicate file '${file}'`);
    seen.add(file);

    const name = (row.name || "").trim();
    if (!name) throw new Error(`tags.csv line ${line} (${file}): empty 'name'`);

    const rawInds = splitList(row.industries);
    const rawTyps = splitList(row.types);
    if (!rawInds.length) throw new Error(`tags.csv line ${line} (${file}): no industries`);
    if (!rawTyps.length) throw new Error(`tags.csv line ${line} (${file}): no types`);

    let industries, types;
    try {
      industries = rawInds.map(normalizeIndustry);
      types = rawTyps.map(normalizeType);
    } catch (e) {
      throw new Error(`tags.csv line ${line} (${file}): ${e.message}`);
    }

    const tier = parseInt(row.tier, 10);
    if (!Number.isInteger(tier) || tier < 1 || tier > 5) {
      throw new Error(`tags.csv line ${line} (${file}): tier must be 1–5, got '${row.tier}'`);
    }

    const yearRaw = (row.year || "").trim();
    const year = yearRaw ? parseInt(yearRaw, 10) : null;

    const activeRaw = (row.active || "").trim().toLowerCase();
    let active = true;
    if (TRUE.has(activeRaw)) active = true;
    else if (FALSE.has(activeRaw)) active = activeRaw === "" ? true : false;
    else throw new Error(`tags.csv line ${line} (${file}): invalid 'active' = '${row.active}'`);

    entries.push({ file, name, industries, types, tier, year, active, notes: row.notes || "" });
  });

  if (!entries.length) throw new Error("tags.csv has no rows");
  return entries;
}

export function activeEntries(entries) {
  return entries.filter((e) => e.active);
}

export function matches(entry, industries, types, matchAll) {
  const checks = [];
  if (industries.length) checks.push(industries.some((i) => entry.industries.includes(i)));
  if (types.length) checks.push(types.some((t) => entry.types.includes(t)));
  if (!checks.length) return true;
  return matchAll ? checks.every(Boolean) : checks.some(Boolean);
}

File: engine/library.py
"""Load and validate a profile's ``tags.csv`` against the controlled vocab.

Each row becomes a :class:`LogoEntry`. Validation is loud by design: an
unknown industry/type, a malformed tier, a duplicate filename, or a missing
master image all raise before any rendering happens. The library is the
source of truth — if it's wrong, we stop here, not three steps downstream.
"""

from __future__ import annotations

import csv
from dataclasses import dataclass, field
from pathlib import Path

from . import vocab

_TRUE = {"true", "1", "yes", "y", "t"}
_FALSE = {"false", "0", "no", "n", "f", ""}

_REQUIRED_COLUMNS = ("file", "name", "industries", "types", "tier")


class LibraryError(ValueError):
    """Raised when the library (folder layout or tags.csv) is invalid."""


@dataclass
class LogoEntry:
    file: str
    name: str
    industries: list[str]
    types: list[str]
    tier: int
    year: int | None = None
    active: bool = True
    notes: str = ""
    path: Path = field(default=Path(), repr=False)

    def matches(self, industries: list[str], types: list[str], match_all: bool) -> bool:
        """True if this entry satisfies the query under the given match mode.

        Only axes the caller actually supplied are tested. ``any`` matches when
        at least one supplied axis hits; ``all`` requires every supplied axis to
        hit. An empty query (no axes) matches everything.
        """
        checks: list[bool] = []
        if industries:
            checks.append(bool(set(industries) & set(self.industries)))
        if types:
            checks.append(bool(set(types) & set(self.types)))
        if not checks:
            return True
        return all(checks) if match_all else any(checks)


@dataclass
class Library:
    profile: str
    root: Path
    logos_dir: Path
    entries: list[LogoEntry]

    def active_entries(self) -> list[LogoEntry]:
        return [e for e in self.entries if e.active]


def _split_list(raw: str) -> list[str]:
    return [part.strip() for part in raw.split("|") if part.strip()]


def _parse_bool(raw: str, *, default: bool, file: str) -> bool:
    v = (raw or "").strip().lower()
    if v in _TRUE:
        return True
    if v in _FALSE:
        return default if v == "" else False
    raise LibraryError(f"{file}: invalid boolean for 'active': {raw!r}")


def _parse_int(raw: str, field_name: str, file: str) -> int:
    try:
        return int(str(raw).strip())
    except (ValueError, TypeError) as exc:
        raise LibraryError(f"{file}: '{field_name}' must be an integer, got {raw!r}") from exc


def load_library(profile_dir: str | Path, *, profile: str | None = None) -> Library:
    """Load and validate the library under ``profile_dir``."""
    root = Path(profile_dir)
    profile = profile or root.name
    tags_path = root / "tags.csv"
    logos_dir = root / "logos"

    if not tags_path.exists():
        raise LibraryError(f"profile {profile!r}: tags.csv not found at {tags_path}")
    if not logos_dir.is_dir():
        raise LibraryError(f"profile {profile!r}: logos/ not found at {logos_dir}")

    with tags_path.open(encoding="utf-8-sig", newline="") as fh:
        reader = csv.DictReader(fh)
        header = reader.fieldnames or []
        missing = [c for c in _REQUIRED_COLUMNS if c not in header]
        if missing:
            raise LibraryError(
                f"tags.csv is missing required column(s): {', '.join(missing)}"
            )

        entries: list[LogoEntry] = []
        seen: set[str] = set()
        for lineno, row in enumerate(reader, start=2):
            file = (row.get("file") or "").strip()
            if not file:
                raise LibraryError(f"tags.csv line {lineno}: empty 'file' value")
            if file in seen:
                raise LibraryError(f"tags.csv line {lineno}: duplicate file {file!r}")
            seen.add(file)

            name = (row.get("name") or "").strip()
            if not name:
                raise LibraryError(f"tags.csv line {lineno} ({file}): empty 'name'")

            raw_inds = _split_list(row.get("industries", ""))
            raw_typs = _split_list(row.get("types", ""))
            if not raw_inds:
                raise LibraryError(f"tags.csv line {lineno} ({file}): no industries")
            if not raw_typs:
                raise LibraryError(f"tags.csv line {lineno} ({file}): no types")

            try:
                industries = [vocab.normalize_industry(v) for v in raw_inds]
                types = [vocab.normalize_type(v) for v in raw_typs]
            except vocab.VocabError as exc:
                raise LibraryError(f"tags.csv line {lineno} ({file}): {exc}") from exc

            tier = _parse_int(row.get("tier", ""), "tier", f"tags.csv line {lineno}")
            if not 1 <= tier <= 5:
                raise LibraryError(
                    f"tags.csv line {lineno} ({file}): tier must be 1–5, got {tier}"
                )

            year_raw = (row.get("year") or "").strip()
            year = _parse_int(year_raw, "year", f"tags.csv line {lineno}") if year_raw else None

            active = _parse_bool(row.get("active", ""), default=True, file=f"tags.csv line {lineno}")

            img_path = logos_dir / file
            if not img_path.exists():
                raise LibraryError(
                    f"tags.csv line {lineno}: master image not found for {file!r} "
                    f"(expected {img_path})"
                )

            entries.append(
                LogoEntry(
                    file=file,
                    name=name,
                    industries=industries,
                    types=types,
                    tier=tier,
                    year=year,
                    active=active,
                    notes=(row.get("notes") or "").strip(),
                    path=img_path,
                )
            )

    if not entries:
        raise LibraryError(f"tags.csv for profile {profile!r} has no rows")

    return Library(profile=profile, root=root, logos_dir=logos_dir, entries=entries)
